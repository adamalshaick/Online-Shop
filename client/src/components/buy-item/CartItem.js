import React, { Component } from "react";
import PropTypes from "prop-types";

class CartItem extends Component {
  render() {
    const { item } = this.props;
    return (
      <div style={{ width: "100%" }}>
        <img
          style={{ height: "50px", width: "50px" }}
          src={`../../uploads/post_image/${item.itemImage}`}
          alt=""
        />
        {item.title}
        <button
          style={{ float: "right" }}
          // onClick={this.onDeleteClick()}
          className="btn btn-danger ml-3"
        >
          Remove
        </button>
        <span style={{ float: "right" }} className="mt-2">
          {item.price} $
        </span>

        <hr />
      </div>
    );
  }
}

CartItem.propTypes = {
  item: PropTypes.object.isRequired
};

export default CartItem;
