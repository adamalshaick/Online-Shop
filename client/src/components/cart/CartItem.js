import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { removeItemFromCart } from "../../actions/cartActions";
import { getProfileById } from "../../actions/profileActions";
import { Link } from "react-router-dom";

export class CartItem extends Component {
  componentDidMount() {
    this.props.getProfileById(this.props.item.user);
  }

  onDeleteClick(id) {
    this.props.removeItemFromCart(id);
  }

  render() {
    const { item, loading } = this.props;
    const { profileById } = this.props.profile;
    return (
      <div style={{ width: "100%" }}>
        <img
          style={{ height: "50px", width: "50px" }}
          // src={`../../uploads/post_image/${item.itemImage}`}
          src="../../uploads/post_image/placeholder.png"
          alt=""
        />
        {item.title}

        <button
          onClick={this.onDeleteClick.bind(this, item._id)}
          type="button"
          className="btn btn-danger ml-3 float-right"
        >
          Remove
        </button>

        {profileById === null || loading ? null : (
          <Link
            to={`/profile/${profileById.handle}`}
            className="btn btn-light float-right ml-3 d-none d-sm-block"
          >
            Seller's Profile
          </Link>
        )}

        <span className="mt-2 float-right">{item.price} $</span>
        <hr />
      </div>
    );
  }
}

CartItem.propTypes = {
  item: PropTypes.object.isRequired,
  removeItemFromCart: PropTypes.func.isRequired,
  profile: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  profile: state.profile
});

export default connect(
  mapStateToProps,
  { removeItemFromCart, getProfileById }
)(CartItem);
