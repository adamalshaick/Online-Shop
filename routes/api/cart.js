const express = require("express");
const router = express.Router();
const mongoose = require("mongoose");
const passport = require("passport");

// Item model
const Item = require("../../models/Item");

// Card model
const Cart = require("../../models/Cart");

// @route   POST api/cart
// @desc    Add item to cart
// @access  Private
router.post(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    // const { errors, isValid } = validateCartInput(req.body);

    // Check Validation
    // if (!isValid) {
    //   // If any errors, send 400 with errors object
    //   return res.status(400).json(errors);
    // }

    Cart.findOne({ user: req.user.id })
      .then(cart => {
        if (cart) {
          cart.items.unshift({ item: req.body });
          cart.value = +card.value + +req.body.price;
          cart.save().then(card => res.json(card));
        } else {
          const cartFields = {
            user: req.user.id,
            items: [{ item: req.body }],
            value: req.body.price
          };

          new Cart(cartFields).save().then(cart => res.json(cart));
        }
      })

      .catch(err => {
        res.json(err);
      });
  }
);

// @route   GET api/cart
// @desc    Get items from cart
// @access  Private
router.get(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    Cart.findOne({ user: req.user.id })
      .then(cart => {
        res.json(cart);
      })
      .catch(err => {
        res.json(err);
      });
  }
);

module.exports = router;
