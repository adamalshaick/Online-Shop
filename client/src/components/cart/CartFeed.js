import React from "react";
import PropTypes from "prop-types";
import CartItem from "./CartItem";
import { Header } from "../common/styles/Header";

export const CartFeed = ({ cartItems }) => {
  let cartValue = 0;
  cartItems.forEach(item => {
    cartValue = +item.price + cartValue;
  });
  return (
    <div>
      <div className="entry">
        <Header>
          <span style={{ fontSize: "18px", float: "left" }}>
            <i className="mr-2 fas fa-shopping-cart" />
            <strong>Your cart</strong>
          </span>
          <span style={{ fontSize: "18px", float: "right" }}>
            <strong>Amount: {cartValue} $</strong>
          </span>
        </Header>

        {cartItems.map(item => (
          <CartItem key={item._id} item={item} />
        ))}
      </div>
    </div>
  );
};

CartFeed.propTypes = {
  cartItems: PropTypes.array.isRequired
};

export default CartFeed;
