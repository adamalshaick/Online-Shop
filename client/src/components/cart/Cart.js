import React from "react";
import fetchCurrentUser from "../common/hoc/fetchCurrentUser";
import CartFeed from "./CartFeed";
import PropTypes from "prop-types";

export const Cart = ({ user }) => {
  return (
    <div className="container">
      <CartFeed cartItems={user.currentUser.cart} />
    </div>
  );
};

Cart.propTypes = {
  user: PropTypes.object.isRequired
};

export default fetchCurrentUser(Cart);
