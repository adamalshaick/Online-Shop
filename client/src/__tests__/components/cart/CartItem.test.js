import React from "react";
import { CartItem } from "../../../components/cart/CartItem";
import { shallow } from "enzyme";

const mockDeletefn = jest.fn();

const Props = {
  item: {}
};

const wrapper = shallow(
  <CartItem {...Props} removeItemFromCart={mockDeletefn} />
);

it("displays a button", () => {
  expect(wrapper.find("button").length).toEqual(1);
});

it("should call the mock delete function", () => {
  wrapper.find("button").simulate("click");
  expect(mockDeletefn.mock.calls.length).toBe(1);
});
