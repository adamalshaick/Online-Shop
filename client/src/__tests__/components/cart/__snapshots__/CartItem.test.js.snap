// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<div
  className="entry2x"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <img
    alt=""
    src="../../uploads/post_image/placeholder.png"
    style={
      Object {
        "height": "50px",
        "width": "50px",
      }
    }
  />
  <button
    className="btn btn-danger ml-3 float-right"
    onClick={[Function]}
    type="button"
  >
    Remove
  </button>
  <Link
    className="btn btn-light float-right ml-3 d-none d-sm-block"
    replace={false}
    to="/profile/undefined"
  >
    Seller's Profile
  </Link>
  <span
    className="mt-2 float-right"
  >
     $
  </span>
  <hr />
</div>
`;
